### DISCUSSION TITLE
XML to JSON conversion using inbuild converters

### DISCUSSION THREAD

**Discussion by vigneshr04** [nextgenhealthcare/connect#5862] [General]
Hi All, im new to Mirth connect.
i am working on a requirement to convert HL7 data to DICOM.
directly i could not find any approach. I tried first creating Hl7 to XML, it is successful.
but XML to DICOM conversion im facing problems.
im using SerializerFactory.getSerializer('DICOM').fromXML(msg);
File is empty always

**Comment by jonbartels** [comment]
Can you post a sample of `msg` as an attachment to this thread?

I am not a DICOM expert. I can not answer your question, but a file will help an expert answer you.

**Comment by vigneshr04** [comment]
Sample XML message:
msg =<HL7Message><MSH><MSH.1>|</MSH.1><MSH.2>^~\&amp;</MSH.2><MSH.3><MSH.3.1>INCEPTO-GATEWAY</MSH.3.1></MSH.3><MSH.4><MSH.4.1>INCEPTO-GATEWAY</MSH.4.1></MSH.4><MSH.5><MSH.5.1>GE-RIS</MSH.5.1></MSH.5><MSH.6><MSH.6.1>GE-RIS</MSH.6.1></MSH.6><MSH.7/><MSH.8/><MSH.9><MSH.9.1>ORU</MSH.9.1><MSH.9.2>R01</MSH.9.2></MSH.9><MSH.10><MSH.10.1>140</MSH.10.1></MSH.10><MSH.11><MSH.11.1>P</MSH.11.1></MSH.11><MSH.12><MSH.12.1>2.7</MSH.12.1></MSH.12><MSH.13/><MSH.14/><MSH.15/><MSH.16/><MSH.17/><MSH.18><MSH.18.1>UNICODE UTF-8</MSH.18.1></MSH.18><MSH.19/><MSH.20/></MSH><PID><PID.1/><PID.2/><PID.3><PID.3.1>DEMO VLN - Polaris4</PID.3.1></PID.3><PID.4/><PID.5><PID.5.1>VLN004</PID.5.1></PID.5><PID.6/><PID.7><PID.7.1>19880101</PID.7.1></PID.7><PID.8><PID.8.1>M</PID.8.1></PID.8><PID.9/></PID><PV1><PV1.1/><PV1.2><PV1.2.1>U</PV1.2.1></PV1.2></PV1><OBR><OBR.1/><OBR.2/><OBR.3><OBR.3.1>ACC_VLN004</OBR.3.1></OBR.3><OBR.4/><OBR.5/><OBR.6/><OBR.7/><OBR.8/><OBR.9/><OBR.10/><OBR.11/><OBR.12/><OBR.13/><OBR.14/><OBR.15/><OBR.16/><OBR.17/><OBR.18/><OBR.19/><OBR.20/><OBR.21><OBR.21.1>ACC_VLN004</OBR.21.1></OBR.21><OBR.22/></OBR><OBX><OBX.1><OBX.1.1>1</OBX.1.1></OBX.1><OBX.2><OBX.2.1>ST</OBX.2.1></OBX.2><OBX.3><OBX.3.1>AI_RESULTS_02</OBX.3.1><OBX.3.2/><OBX.3.3>INCEPTO</OBX.3.3></OBX.3><OBX.4/><OBX.5><OBX.5.1>FINDINGS</OBX.5.1><OBX.5.2>LUNG NODULE DETECTION AND MEASUREMENTS</OBX.5.2></OBX.5><OBX.6/><OBX.7/><OBX.8/><OBX.9/><OBX.10/><OBX.11><OBX.11.1>R</OBX.11.1></OBX.11><OBX.12/><OBX.13/><OBX.14><OBX.14.1>20220525130115</OBX.14.1></OBX.14><OBX.15><OBX.15.1/><OBX.15.2/></OBX.15><OBX.16/><OBX.17><OBX.17.1>AI_analysis</OBX.17.1></OBX.17><OBX.18/><OBX.19/><OBX.20/></OBX><NTE><NTE.1><NTE.1.1>1</NTE.1.1></NTE.1><NTE.2><NTE.2.1>Polaris</NTE.2.1></NTE.2><NTE.3><NTE.3.1>https://www.polaris.incep.to/analyses/&lt;id&gt;</NTE.3.1></NTE.3><NTE.4/><NTE.5/><NTE.6/><NTE.7/><NTE.8/><NTE.9/><NTE.10/><NTE.11/><NTE.12/><NTE.13/><NTE.14/><NTE.15/><NTE.16/><NTE.17/></NTE></HL7Message>

i want to covert to DICOM in destination and send in DICOM node.
While parsing i am getting parsing error but could not find what ?

I'm using transformer as well:
import Package(Packages.org.dcm4che2.data);

// convert template to a dcm4che DicomObject
var dicomBytes = FileUtil.decode(SerializerFactory.getSerializer('DICOM').fromXML(tmp));
var DICOMMESSAGE = DICOMUtil.byteArrayToDicomObject(dicomBytes, false);

channelMap.put("mess", DICOMMESSAGE);


Error Message : WARNING: Failed to parse C:\Program Files\Mirth Connect\appdata\temp\temp7350631049288511612tmp - skipped.

**Reply by jonbartels** [reply]
That XML is the MC representation of an HL7 message. You could certainly call `SerializerFactory.getSerializer('HL7').fromXML(msg)` and get good output but to get DICOM you have to map your HL7 input to a DICOM template, then you can serialize that DICOM and send it to another system

**Reply by vigneshr04** [reply]
Yeah , right now im doing that way , for now i have hard coded values in Dicom object and mapping few mandatory fields
**below is the code for transformer:**
importPackage(Packages.org.dcm4che2.data);

// prerequisite: have a skeleton with all the static DICOM parts stored in the outbound message template 'tmp'

// convert template to a dcm4che DicomObject
var dicomBytes = SerializerFactory.getSerializer('DICOM').fromXML(tmp);
var dcmObject = DICOMUtil.byteArrayToDicomObject(dicomBytes, false);

// miscellaneous setup
var random = new java.util.Random();
var now = new java.util.Date();

//  UID prefixes
var UID_PREFIX = "9.8.7.6.5."; // get your own unique UID root to use here
var STUDY_SUB = "1.";
var SERIES_SUB = "2.";
var INSTANCE_SUB = "3.";


dcmObject.putString(Tag.MediaStorageSOPInstanceUID, VR.UI, "1.3.12.2.1107.5.8.21.22202125184636021332314634160421156025");
dcmObject.putString(Tag.SOPInstanceUID, VR.UI, "1.3.12.2.1107.5.8.21.22202125184636021332314634160421156025");

// patient demographics from HL7
dcmObject.putString(Tag.PatientName, VR.PN,msg['MSH']['MSH.18']['MSH.18.1'].toString() );
dcmObject.putString(Tag.PatientID, VR.LO, msg['MSH']['MSH.18']['MSH.18.1'].toString());
dcmObject.putString(Tag.PatientBirthDate, VR.DA, msg['MSH']['MSH.18']['MSH.18.1'].toString());
dcmObject.putString(Tag.PatientSex, VR.CS,msg['MSH']['MSH.18']['MSH.18.1'].toString());

dcmObject.putString(Tag.StudyInstanceUID, VR.UI, "1.3.12.2.1107.5.8.21.161766888302691154657029524413728644541");

dcmObject.putString(Tag.SeriesInstanceUID, VR.UI, "1.3.12.2.1107.5.8.21.214919502621970086568010755839221524033");

dcmObject.putInt(Tag.SeriesNumber, VR.IS, 99);

dcmObject.putString(Tag.SeriesDescription, VR.LO, "RaasTest");

tmp = SerializerFactory.getSerializer('DICOM').toXML( FileUtil.encode(DICOMUtil.dicomObjectToByteArray(dcmObject)));

console.warn(tmp);

channelMap.put("mess", dcmObject);.


**Code looks fine to me, but im getting below error:**
Transformer error
ERROR MESSAGE: Error evaluating transformer
com.mirth.connect.server.MirthJavascriptTransformerException: 
CHANNEL:	xmltodicom
CONNECTOR:	Destination 1
SCRIPT SOURCE:	TRANSFORMER
SOURCE CODE:	
106: 
107: // prerequisite: have a skeleton with all the static DICOM parts stored in the outbound message template 'tmp'
108: 
109: // convert template to a dcm4che DicomObject
110: var dicomBytes = SerializerFactory.getSerializer('DICOM').fromXML(tmp);
111: var dcmObject = DICOMUtil.byteArrayToDicomObject(dicomBytes, false);
112: 
113: // miscellaneous setup
114: var random = new java.util.Random();
115: var now = new java.util.Date();
LINE NUMBER:	111
DETAILS:	Can't find method com.mirth.connect.server.userutil.DICOMUtil.byteArrayToDicomObject(java.lang.String,boolean).
	at 2655de4a-8027-4465-bfe8-0c2b215dae00_JavaScript_Filter_Transformer_1:111 (doTransform)
	at 2655de4a-8027-4465-bfe8-0c2b215dae00_JavaScript_Filter_Transformer_1:167 (doScript)
	at 2655de4a-8027-4465-bfe8-0c2b215dae00_JavaScript_Filter_Transformer_1:169
	at com.mirth.connect.server.transformers.JavaScriptFilterTransformer$FilterTransformerTask.doCall(JavaScriptFilterTransformer.java:235)
	at com.mirth.connect.server.transformers.JavaScriptFilterTransformer$FilterTransformerTask.doCall(JavaScriptFilterTransformer.java:187)
	at com.mirth.connect.server.util.javascript.JavaScriptTask.call(JavaScriptTask.java:114)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:750)

do you know the reason for this?

**Comment by jtyme** [comment]
I am no expert but just looking at the error details and then looking at lines 111 and 110 it looks like dicomeBytes is a string representation and not a byteArray. From looking here: http://javadocs.mirthcorp.com/connect/3.0.0/user-api/com/mirth/connect/server/userutil/DICOMUtil.html I am not seeing a byteArrayToDicomObject that takes just a java string and boolean value.  Think you may need to convert dicomBytes to a byte array and then pass it into the function.

**Reply by ab-20-m** [reply]
I think you nailed it. `DICOMSerializer.fromXML(string)` returns a [Base64 encoded string](https://github.com/nextgenhealthcare/connect/blob/010f2a4dc636ad1c24b20a7fd1f625c5f6b91a39/server/src/com/mirth/connect/plugins/datatypes/dicom/DICOMSerializer.java#L143) of the DICOM object bytes, not the raw bytes of the object.

**Reply by vigneshr04** [reply]
You guys are right, it worked for me after converting string into byte array, Thank you very much.
